File: backend/test_code.js
// Vulnerable JavaScript Code for Testing VulnVault
// This file contains multiple security vulnerabilities

// 1. Hardcoded credentials
const API_KEY = "sk-1234567890abcdef";
const DB_PASSWORD = "admin123";
const SECRET_TOKEN = "super-secret-token";

// 2. SQL Injection vulnerability
function getUserById(userId) {
    const query = "SELECT * FROM users WHERE id = " + userId;
    db.query(query);  // Dangerous!
}

// 3. XSS vulnerability
function displayUserInput(userInput) {
    document.getElementById("output").innerHTML = userInput;  // XSS risk
    document.write(userInput);  // Also dangerous
}

// 4. Command Injection
const exec = require('child_process').exec;

function pingHost(hostname) {
    exec('ping -c 4 ' + hostname);  // Command injection!
}

// 5. Unsafe eval
function calculateExpression(expression) {
    return eval(expression);  // Never use eval!
}

// 6. Insecure randomness
function generateToken() {
    return Math.random().toString(36);  // Not cryptographically secure
}

// 7. Insecure cookie
document.cookie = "session=" + sessionId;  // Should use secure, httpOnly flags

// 8. Weak cryptography
const crypto = require('crypto');
function hashPassword(password) {
    return crypto.createHash('md5').update(password).digest('hex');  // MD5 is broken!
}

// 9. Path traversal
const fs = require('fs');
function readFile(filename) {
    return fs.readFileSync('/uploads/' + filename);  // Path traversal risk
}

// 10. Prototype pollution
function merge(target, source) {
    for (let key in source) {
        target[key] = source[key];  // Vulnerable to prototype pollution
    }
}

// 11. RegEx DoS
function validateEmail(email) {
    const regex = /^([a-zA-Z0-9])(([a-zA-Z0-9])*([\._-])?([a-zA-Z0-9]))*@([a-zA-Z0-9])+((\.)([a-zA-Z0-9])+)+$/;
    return regex.test(email);  // Can cause DoS with crafted input
}

console.log("This file contains 11+ vulnerabilities for testing!");

File: backend/test_code.py
"""
Intentionally vulnerable Python code for testing VulnVault
This file contains common security vulnerabilities
"""
import os
import pickle
import subprocess

# Vulnerability 1: Hardcoded password
PASSWORD = "admin123"
API_KEY = "sk-1234567890abcdef"

# Vulnerability 2: SQL Injection
def get_user(user_id):
    import sqlite3
    conn = sqlite3.connect('database.db')
    cursor = conn.cursor()
    # BAD: String formatting in SQL query
    query = f"SELECT * FROM users WHERE id = {user_id}"
    cursor.execute(query)
    return cursor.fetchone()

# Vulnerability 3: Command Injection
def run_command(user_input):
    # BAD: Direct execution of user input
    os.system(user_input)
    
# Vulnerability 4: Unsafe Deserialization
def load_data(filename):
    with open(filename, 'rb') as f:
        # BAD: Pickle can execute arbitrary code
        return pickle.load(f)

# Vulnerability 5: Path Traversal
def read_file(filename):
    # BAD: No validation of filename
    with open(f"/var/data/{filename}", 'r') as f:
        return f.read()

# Vulnerability 6: Shell Injection
def ping_host(host):
    # BAD: Using shell=True with user input
    subprocess.call(f"ping -c 1 {host}", shell=True)

# Vulnerability 7: Weak cryptography
def encrypt_password(password):
    import hashlib
    # BAD: MD5 is cryptographically broken
    return hashlib.md5(password.encode()).hexdigest()

# Vulnerability 8: Using exec()
def run_user_code(code):
    # BAD: exec() can run arbitrary code
    exec(code)

# Vulnerability 9: Assert used for security
def is_admin(user):
    # BAD: Asserts can be disabled
    assert user.is_admin == True
    return True

# Vulnerability 10: Insecure random
import random
def generate_token():
    # BAD: random is not cryptographically secure
    return random.randint(1000, 9999)

if __name__ == "__main__":
    print("This file contains 10+ security vulnerabilities!")
    print("VulnVault should detect most of them.")
